packages/@aws-cdk/aws-dynamodb/lib/table.ts

```diff
@@ -262,18 +297,74 @@ export class Table extends Construct {
     this.writeScalingPolicyResource = this.buildAutoScaling(this.writeScalingPolicyResource, 'Write', props);
   }
 
+  /**
+   * Validate the table construct.
+   *
+   * @returns an array of validation error message
+   */
   public validate(): string[] {
     const errors = new Array<string>();
-    if (!this.findKey(HASH_KEY_TYPE)) {
+
+    const tablePartitionKey = this.findKey(HASH_KEY_TYPE);
+    if (!tablePartitionKey) {
       errors.push('a partition key must be specified');
     }
+    if (this.localSecondaryIndexes.length > 0) {
+      if (!this.findKey(RANGE_KEY_TYPE)) {
+        // missing a sort key for the table causes validation exception, status code 400, while trying to create CFN stack
+        errors.push('a sort key of the table must be specified to add local secondary indexes');
+      }
+      if (tablePartitionKey && tablePartitionKey.attributeName !== this.localSecondaryIndexPartitionKeyName) {
+        // https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/LSI.html
+        errors.push('a partition key of local secondary index must be the same as that of the table');
```

Comment:
If we don't specify a partition key, then we could not infer the second scenario in the below.

```
// OK
table.addPartitionKey();
table.addLocalSecondaryIndex();

// we cannot infer a partition key of table when addLocalSecondaryIndex() is called.
table.addLocalSecondaryIndex();
table.addPartitionKey();
```

All validation complexities come from the fact that we allow lazy initialization of a base table's key schema via `addPartitionKey`. An underlying principle of the PR is that all public methods should be independent each other.

Reply:
I see. That's an admirable goal! 

Still, I think we'd also be perfectly fine with requiring that people call `addPartitionKey()` before calling `addLocalSecondaryIndex()` (so we can just check for it and throw an error), thereby simplifying the implementation a lot.

If you insist on either call order, you can also use `cdk.Token()` (which lazily evaluates values) to return the eventual partition key during synthesis. You'll have to change the implementation somewhat, but it should end up looking like:

```ts
this.localSecondaryIndexes.push({
  keySchema: [
    new cdk.Token(() => this.partitionKeySchema),  // <-- notice this
    rangeKeySchema
  ],
  indexName: ...
  projection: ...
});
```

What the `cdk.Token` does is delay the retrieval of `this.partitionKeySchema` until after the complete setup has happened, so that it has a chance to have been set later on.

Reply:
> Still, I think we'd also be perfectly fine with requiring that people call `addPartitionKey()` before calling `addLocalSecondaryIndex()` (so we can just check for it and throw an error), thereby simplifying the implementation a lot.

I will go this way. I agree with the points you mentioned. In addition, not adding a partition key to `LocalSecondaryIndexProps` would be more aligned with the LSI nature, the same partition key as the one of a base table.